src/lex.js

```diff
@@ -567,92 +590,108 @@ Lexer.prototype = {
+        if (!state.inESNext(true)) {
+          this.trigger("warning", {
+            code: "W119",
+            line: this.line,
+            character: this.char + index,
+            data: [ "\\u{...}" ]
+          });
         }
-
-        return null;
-      }
-
-      return null;
-    }.bind(this);
-
-    var getIdentifierStart = function() {
-      /*jshint validthis:true */
-      var chr = this.peek(index);
-      var code = chr.charCodeAt(0);
-
-      if (code === 92) {
-        return readUnicodeEscapeSequence();
-      }
-
-      if (code < 128) {
-        if (asciiIdentifierStartTable[code]) {
-          index += 1;
-          return chr;
+        escapeSequence += "{";
+        index += 3;
+        var i;
+        for (i = 0; i < 7; i++, index++) {
+          var ch = this.peek(index);
+          if (ch === "}") {
+            escapeSequence += "}";
+            break;
+          } else if (i < 6 && isHexNumber(ch)) {
+            hexCode += ch;
+            escapeSequence += ch;
+          } else {
+            this.trigger("error", {
+              code: "E020",
+              line: this.line,
+              character: this.char + index,
+              data: [ "}", "{", this.line, ch ]
+            });
+            return escapeSequence;
+          }
         }
-
-        return null;
+        index += 1;
+      } else { // \uXXXX
+        hexCode = this.input.substr(index + 2, 4);
+        escapeSequence += hexCode;
+        index += 6;
       }
 
-      if (isNonAsciiIdentifierStart(code)) {
-        index += 1;
-        return chr;
+      var code = parseInt(hexCode, 16);
+      if (!isHexNumber(hexCode) || !isValidIdentifierCharacter(code, position)) {
+        this.trigger("warning", {
+          code: "W052",
+          line: this.line,
+          character: this.char + index - escapeSequence.length,
+          data: [ escapeSequence ]
+        });
       }
 
-      return null;
+      return escapeSequence;
     }.bind(this);
 
-    var getIdentifierPart = function() {
+    var getIdentifier = function(position) {
       /*jshint validthis:true */
       var chr = this.peek(index);
-      var code = chr.charCodeAt(0);
+      var code = codePoint(chr);
 
       if (code === 92) {
-        return readUnicodeEscapeSequence();
+        return readUnicodeEscapeSequence(position);
       }
 
-      if (code < 128) {
-        if (asciiIdentifierPartTable[code]) {
-          index += 1;
-          return chr;
+      if (0xd800 <= code && code <= 0xdbff) { // high surrogate
+        var ch2 = this.peek(index + 1);
+        var cp = codePoint(ch2);
+        if (0xdc00 <= cp && cp <= 0xdfff) { // low surrogate
+          chr += ch2;
+          code = codePoint(chr);
+          if (!state.inESNext(true)) {
+            this.trigger("warning", {
+              code: "W119",
+              line: this.line,
+              character: this.char + index,
+              data: [ "Astral symbols in identifiers" ]
```

Comment:
How should the warning message be?
`'Astral symbols in identifiers' is only available in ES6 (use esnext option).` isn't obviously ok. (`symbols`...`is`)